### PR TITLE
[KAT-538] Encapsulate node and edge properties

### PR DESCRIPTION
**Pull request by witchel**
This change encapsulates the node and edge property tables for our graphs.
Exposing the arrow table of properties makes it impossible to change our
representation to keep a subset of properties in memory at once, so we
stop doing that.

There is a dependent PR (https://github.com/KatanaGraph/katana-enterprise/pull/1479) for enterprise that has more detail.

### CODE REVIEW

File: libgalois/include/katana/PropertyGraph.h [hunk 1]
```diff
@@ -262,38 +262,51 @@ class KATANA_EXPORT PropertyGraph {
...
     }
 
-    std::vector<std::string> property_names() const {
-      return properties()->ColumnNames();
+    std::shared_ptr<arrow::ChunkedArray> GetProperty(
+        const std::string& str) const {
+      return (const_g->*property_fn_str)(str);
     }
 
+    int32_t GetPropertyNum() const { return (const_g->*property_num_fn)(); }
```
Comment (ddn0) [inline libgalois/include/katana/PropertyGraph.h]: idle comment: every time I read this, I always think "shouldn't this be called GetNumProperties", but I think the last time this came up, we decided the current `GetFooNum` was consistent with the current API so we have kept things as there are now (cc @roshandathathri who is the person I strongly associate with this discussion)

**Review by ddn0 — APPROVED** [review]

File: libgalois/include/katana/PropertyGraph.h (hunk 1 above)
Comment (witchel) [reply libgalois/include/katana/PropertyGraph.h]: I have updated the names to `GetNumProperties()`.

File: libgalois/include/katana/PropertyGraph.h [hunk 2]
```diff
@@ -262,38 +262,51 @@ class KATANA_EXPORT PropertyGraph {
...
+  /// Return the node property table for local nodes
+  const std::shared_ptr<arrow::Table>& node_properties() const {
+    return rdg_.node_properties();
+  }
+
+  /// Return the edge property table for local edges
+  const std::shared_ptr<arrow::Table>& edge_properties() const {
+    return rdg_.edge_properties();
+  }
+
```
Comment (arthurp) [inline libgalois/include/katana/PropertyGraph.h:275]: Do we want to expose this? Will we always be able to construct an arrow table regardless of the internal storage of the properties?

File: libgalois/include/katana/PropertyGraph.h [hunk 3]
```diff
@@ -262,38 +262,51 @@ class KATANA_EXPORT PropertyGraph {
...
 public:
   /// PropertyView provides a uniform interface when you don't need to
   /// distinguish operating on edge or node properties
   struct PropertyView {
     PropertyGraph* g;
+    const PropertyGraph* const_g;
```
Comment (arthurp) [inline libgalois/include/katana/PropertyGraph.h:280]: Why is this needed? I suspect there is a detail of C++ typing I don't understand.

File: libgalois/include/katana/PropertyGraph.h [hunk 4]
```diff
@@ -646,36 +660,59 @@ class KATANA_EXPORT PropertyGraph {
...
         .schema_fn = &PropertyGraph::node_schema,
-        .property_fn = &PropertyGraph::GetNodeProperty,
-        .properties_fn = &PropertyGraph::node_properties,
+        .property_fn_int = &PropertyGraph::GetNodeProperty,
+        .property_fn_str = &PropertyGraph::GetNodeProperty,
+        .property_num_fn = &PropertyGraph::GetNumNodeProperties,
         .add_properties_fn = &PropertyGraph::AddNodeProperties,
         .upsert_properties_fn = &PropertyGraph::UpsertNodeProperties,
         .remove_property_int = &PropertyGraph::RemoveNodeProperty,
         .remove_property_str = &PropertyGraph::RemoveNodeProperty,
```
Comment (arthurp) [inline libgalois/include/katana/PropertyGraph.h]: Why are we using function pointers here instead of a simple subclass? It would eliminate the need for the explicit implementations of all the silly little methods that really just implement a vtable.

File: libgalois/include/katana/PropertyGraph.h [hunk 5]
```diff
@@ -767,6 +804,9 @@ CreateSymmetricGraph(PropertyGraph* pg);
 KATANA_EXPORT Result<std::unique_ptr<PropertyGraph>>
 CreateTransposeGraphTopology(const GraphTopology& topology);
 
+/// Report approximate memory use for a property view
+KATANA_EXPORT uint64_t ApproxMemUse(const PropertyGraph::PropertyView& pview);
+
```
Comment (arthurp) [inline libgalois/include/katana/PropertyGraph.h]: Why not make this a method on PropertyView.

File: libgalois/src/PropertyViews.cpp [hunk 6]
```diff
@@ -19,3 +19,24 @@ katana::internal::ExtractArrays(
...
+  std::vector<arrow::Array*> ret;
+  for (auto& property : properties) {
+    auto column = pview.GetProperty(property);
+    if (!column) {
+      return ErrorCode::PropertyNotFound;
+    }
+    if (column->num_chunks() != 1) {
+      // Katana form graphs only contain single chunk property columns.
+      return ErrorCode::TODO;
+      // TODO: Maybe we need an InvalidGraph error
```
Comment (arthurp) [inline libgalois/src/PropertyViews.cpp]: Username? Also add message to error code?

**Review by arthurp — APPROVED** [review]
Overall, this is OK. I feel like it increases the interface surface more than is required for the features you implement. I would encourage you to reduce the interface if possible. One example, is that you expose both `arrow::Table` and `PropertyView`. They do similar things. Maybe just expose one of them.

File: libgalois/include/katana/PropertyGraph.h (hunk 2 above)
Comment (witchel) [reply libgalois/include/katana/PropertyGraph.h:275]: I'm not following.  I moved these functions from public to private to stop exposing them.

File: libgalois/include/katana/PropertyGraph.h (hunk 3 above)
Comment (witchel) [reply libgalois/include/katana/PropertyGraph.h:280]: Sorry, I asked about this in #1479, which is the associated enterprise PR.  Here are the questions.

In PropertyGraph.h, to implement const and non-const PropertyViews I have a PropertyGraph* called g and a const PropertyGraph* called const_g. Is there a better way of doing this?

For the const PropertyView, I leave several function pointers as null. I could make functions that assert, but it seemed annoying to have several different types of dummy functions that do nothing but assert.

File: libgalois/include/katana/PropertyGraph.h (hunk 2 above)
Comment (arthurp) [reply libgalois/include/katana/PropertyGraph.h:275]: My bad. I wasn't paying attention. This may also invalidate my global comment.

File: libgalois/include/katana/PropertyGraph.h (hunk 4 above)
Comment (witchel) [reply libgalois/include/katana/PropertyGraph.h]: Donald did the PropertyView class initially.  Do you mean a nested class or subclass?  `PropertyView` isn't a subclass of any class.

File: libgalois/include/katana/PropertyGraph.h (hunk 3 above)
Comment (arthurp) [reply libgalois/include/katana/PropertyGraph.h:280]: I see no connection between the two pointer types and the nulls vs asserts. But it's no big deal.

File: libgalois/include/katana/PropertyGraph.h (hunk 4 above)
Comment (arthurp) [reply libgalois/include/katana/PropertyGraph.h]: What I mean is that you could have a pure virtual base class and then a few subclasses implementing that. Not that there is currently such a structure. That said, I just realized why this approach was probably chosen. The methods are mostly just methods that already exist on `PropertyGraph`, so this MIGHT be slightly simpler to implement.

I'll move on. :-)

File: libgalois/include/katana/PropertyGraph.h (hunk 5 above)
Comment (witchel) [reply libgalois/include/katana/PropertyGraph.h]: Oh, good idea, thanks.

File: libgalois/src/PropertyViews.cpp (hunk 6 above)
Comment (witchel) [reply libgalois/src/PropertyViews.cpp]: Ok.  For the record, I copied this code from the function above.  I guess that author did not provide a username for the TODO or an error message.

File: libgalois/include/katana/PropertyGraph.h (hunk 3 above)
Comment (witchel) [reply libgalois/include/katana/PropertyGraph.h:280]: I really am interested if there is a better way to do what I want without `g` and `const_g`.

I included the second question because it seemed like the kind of thing you might know or have an opinion on and I didn't want to make you read #1479.

Comment (ddn0) [reply libgalois/include/katana/PropertyGraph.h:280]: `g` and `const_g` point to the same object? and `const_cast` doesn't work or a function that wraps `const_cast`?

Comment (witchel) [reply libgalois/include/katana/PropertyGraph.h:280]: They point to the same object.
Can you explain your question about `const_cast`?  The problem comes from creating a `PropertyView `and a `const PropertyView`.

Comment (ddn0) [reply libgalois/include/katana/PropertyGraph.h:280]: I see.

One pattern would be to use a template that captures the const or lack of const-ness, e.g.,

```cpp
using ConstView = View<const PropertyGraph*>;
using MutableView = View<PropertyGraph*>;
```

Though given that the const view actually has ~less~ fewer functions (the nullptr member functions), I'd suggest just making two classes and skip the templates, e.g.,
```cpp
class ReadOnlyView {

};

class MutableView: public ReadOnlyView {
public:
  Result<void> AddProperties(....) {....}
};

ReadOnlyView edge_property_view() const {....}
MutableView edge_property_view() {...}
```

whether you want to make `Mutable` and `ReadOnlyView` virtual base classes and have concrete classes for nodes and edges or want to continue using member function pointers is up to you.

I originally used member function pointers because I saw (and still see) PropertyGraph as a closed ecosystem, so how we implemented the required (limited) polymorphism was an implementation detail, so I picked what I thought was the more compact way. Virtual functions are fine too.

Comment (witchel) [reply libgalois/include/katana/PropertyGraph.h:280]: I have taken your advice and made two explicitly named structs.  I included the read-only version in the mutable version because it led to less duplication.  

I introduced explicit names for the "constructors" instead of const/non-const because I found myself casting a `DistPropertyGraph `pointer to a const pointer and invoking `edge_property_view()` which seemed like an unnecessarily obscure mechanism.

Comment (witchel) [reply libgalois/include/katana/PropertyGraph.h:280]: Thank you!
